### PR TITLE
fix(auth-plugins): auth plugins cache inconsistency while anonymous was configured as username

### PR DESCRIPTION
**Pull request by vm-001**
### Summary

As Kong auth plugins support `conifg.anonymous` to be configured as the username of consumer, once a consumer has been cached with the username as part of the cache key, this consumer cache won't be invalidated by the default cache invalidation event handler after the consumer has been deleted.

Considers of the conifg.anonymous exists in several auth plugins, so I added a common event handler to invalidate the extra cache rather than in each plugin.

### Issue reference
FTI-4287

### CODE REVIEW

**Comment by bungle** [comment]
This can be optimized (the Kong EE OpenID Connect plugin has an optimized version of this)

**Comment by vm-001** [comment]
> This can be optimized (the Kong EE OpenID Connect plugin has an optimized version of this)

Nice. https://github.com/Kong/kong/pull/9371/commits/d9a740ae9748ffc1f19db625178aa10b9fce5861

File: kong/runloop/handler.lua [hunk 1]
```diff
@@ -526,6 +527,32 @@ local function register_events()
...
+  -- so add an event handler to invalidate the extra cache in case of data inconsistency
+  worker_events.register(function(data)
+    workspaces.set_workspace(data.workspace)
+
+    local old_entity = data.old_entity
+    local old_username
+    if old_entity then
+      old_username = old_entity.username
+      if old_username and old_username ~= null and old_username ~= "" then
+        kong.cache:invalidate(kong.db.consumers:cache_key(old_entity.username))
```
Comment (chronolaw) [inline kong/runloop/handler.lua:543]: here we should use cached var `old_username` directly.

File: kong/runloop/handler.lua [hunk 2]
```diff
@@ -526,6 +527,32 @@ local function register_events()
...
+      end
+    end
+
+    local entity = data.entity
+    if entity then
+      local username = entity.username
+      if username and username ~= null and username ~= "" and username ~= old_username then
+        kong.cache:invalidate(kong.db.consumers:cache_key(entity.username))
```
Comment (chronolaw) [inline kong/runloop/handler.lua:551]: Same as above, use `username`.

File: spec/03-plugins/25-oauth2/03-access_spec.lua [hunk 3]
```diff
@@ -4067,6 +4067,85 @@ describe("Plugin: oauth2 [#" .. strategy .. "]", function()
...
+    local admin_client
+    local anonymous
+
+    lazy_setup(function()
+      anonymous = assert(admin_api.consumers:insert {
+        username = "test",
+      })
+
+      local service = assert(admin_api.services:insert({
+        url = "https://mockbin.org"
```
Comment (chronolaw) [inline spec/03-plugins/25-oauth2/03-access_spec.lua:4082]: need a `,` at the end of the line.

File: spec/03-plugins/25-oauth2/03-access_spec.lua [hunk 4]
```diff
@@ -4067,6 +4067,85 @@ describe("Plugin: oauth2 [#" .. strategy .. "]", function()
...
+      }))
+
+      local route = assert(admin_api.routes:insert({
+        hosts      = { "anonymous-with-username.com" },
+        protocols  = { "http", "https" },
+        service    = service
```
Comment (chronolaw) [inline spec/03-plugins/25-oauth2/03-access_spec.lua:4088]: need a `,` at the end of the line.

File: spec/03-plugins/25-oauth2/03-access_spec.lua [hunk 5]
```diff
@@ -4067,6 +4067,85 @@ describe("Plugin: oauth2 [#" .. strategy .. "]", function()
...
+      }))
+    end)
+
+    before_each(function()
+      proxy_client = helpers.proxy_client()
+      admin_client = helpers.admin_client()
+    end)
+
+    after_each(function()
+      if proxy_client then proxy_client:close() end
```
Comment (chronolaw) [inline spec/03-plugins/25-oauth2/03-access_spec.lua:4105]: Not recommended, we should not write `if-then` in one line.

File: spec/03-plugins/25-oauth2/03-access_spec.lua [hunk 6]
```diff
@@ -4067,6 +4067,85 @@ describe("Plugin: oauth2 [#" .. strategy .. "]", function()
...
+        local res = assert(proxy_client:send {
+          method = "GET",
+          path = "/request",
+          headers = {
+            ["Host"] = "anonymous-with-username.com",
+          }
+        })
+        assert.response(res).has.status(200)
+        local body = assert.response(res).has.jsonbody()
+        assert.are.equal("true", body.headers["x-anonymous-consumer"])
```
Comment (chronolaw) [inline spec/03-plugins/25-oauth2/03-access_spec.lua:4127]: Should we use `assert.equal`?

File: kong/runloop/handler.lua (hunk 1 above)
Comment (vm-001) [reply kong/runloop/handler.lua:543]: Nice catch.

File: spec/03-plugins/25-oauth2/03-access_spec.lua (hunk 6 above)
Comment (vm-001) [reply spec/03-plugins/25-oauth2/03-access_spec.lua]: Maybe both are fine? I see a lot of usages of `assert.are.equal`

**Review by bungle — APPROVED** [review]

**Review by chronolaw — APPROVED** [review]

**Pull request merged**